code/datums/weather/weather.dm

```diff
@@ -0,0 +1,134 @@
+	var/target_z = ZLEVEL_STATION //The z-level to affect
+
+	var/overlay_layer = 10 //Since it's above everything else, this is the layer used by default. 2 is below mobs and walls if you need to use that.
+	var/aesthetic = FALSE //If the weather has no purpose other than looks
+	var/immunity_type = "storm" //Used by mobs to prevent them from being affected by the weather
+
+	var/stage = END_STAGE //The stage of the weather, from 1-4
+
+	var/probability = FALSE //Percent chance to happen if there are other possible weathers on the z-level
+
+/datum/weather/New()
+	..()
+	weather_master.existing_weather |= src
+
+/datum/weather/Destroy()
+	weather_master.existing_weather -= src
+	return ..()
+
+/datum/weather/proc/telegraph()
+	if(stage == STARTUP_STAGE)
+		return
+	stage = STARTUP_STAGE
+	for(var/V in get_areas(area_type))
+		var/area/A = V
+		if(A.z == target_z)
+			impacted_areas |= A
+	weather_duration = rand(weather_duration_lower, weather_duration_upper)
+	update_areas()
+	for(var/V in player_list)
+		var/mob/M = V
+		if(M.z == target_z)
+			if(telegraph_message)
+				to_chat(M, telegraph_message)
+			if(telegraph_sound)
+				M << sound(telegraph_sound)
+	addtimer(src, "start", telegraph_duration)
+
+/datum/weather/proc/start()
+	if(stage >= MAIN_STAGE)
+		return
+	stage = MAIN_STAGE
+	update_areas()
+	for(var/V in player_list)
+		var/mob/M = V
+		if(M.z == target_z)
+			if(weather_message)
+				to_chat(M, weather_message)
+			if(weather_sound)
+				M << sound(weather_sound)
+	weather_master.processing_weather |= src
+	addtimer(src, "wind_down", weather_duration)
+
+/datum/weather/proc/wind_down()
+	if(stage >= WIND_DOWN_STAGE)
+		return
+	stage = WIND_DOWN_STAGE
+	update_areas()
+	for(var/V in player_list)
+		var/mob/M = V
+		if(M.z == target_z)
+			if(end_message)
+				to_chat(M, end_message)
+			if(end_sound)
+				M << sound(end_sound)
+	weather_master.processing_weather -= src
+	addtimer(src, "end", end_duration)
+
+/datum/weather/proc/end()
+	if(stage == END_STAGE)
+		return
+	stage = END_STAGE
+	update_areas()
+
+/datum/weather/proc/can_impact(mob/living/L) //Can this weather impact a mob?
+	if(L.z != target_z)
+		return
+	if(immunity_type in L.weather_immunities)
+		return
+	if(!(get_area(L) in impacted_areas))
+		return
+	return 1
+
+/datum/weather/proc/impact(mob/living/L) //What effect does this weather have on the hapless mob?
+	return
+
+/datum/weather/proc/update_areas()
+	for(var/V in impacted_areas)
+		var/area/N = V
+		N.layer = overlay_layer
+		N.icon = 'icons/effects/weather_effects.dmi'
+		N.invisibility = 0
+		switch(stage)
+			if(STARTUP_STAGE)
+				N.icon_state = telegraph_overlay
+			if(MAIN_STAGE)
+				N.icon_state = weather_overlay
+			if(WIND_DOWN_STAGE)
+				N.icon_state = end_overlay
+			if(END_STAGE)
+				N.icon_state = initial(N.icon_state)
+				N.icon = 'icons/turf/areas.dmi'
+				N.layer = 10 //Just default back to normal area stuff since I assume setting a var is faster than initial
```

Comment:
Actually not, you see....AREA_LAYER is actually a BYOND define, and they have it defined as 1, by default, so we can't actually use that define....so this is what we'll have to do, for now.

Eventually we should probably do this: https://github.com/tgstation/tgstation/pull/17949 so it's more sane.